### PR TITLE
Fix and narrow backwards compatibility in Travertino

### PR DESCRIPTION
**Pull request by HalfWhitt**
#3679 highlighted the fact that Travertino's shim for accommodating Toga < 0.5's now-broken style application was casting too wide a net; if *any* exception occurred during applicator assignment, it was assumed to be due to a Toga version mismatch.

In the course of setting out to narrow the shim, I realized that the current state of Travertino backwards compatibility is in fact broken, and apparently has been ever since the move to 0.5.0, because it doesn't sufficiently handle the changes to the `apply()` signature from #3159 and #3160. I don't remember exactly how I was testing at the time, but I'm baffled at how I managed to miss that. 😳 It's frustrating to know it's been broken all this time, despite all the work put into accommodations, and I'm honestly surprised no one's come asking about it. (I guess anyone who's run up against it has just seen the deprecation messages and updated/pinned.)

For posterity, here are the two best ways I've figured out so far to test this backwards compatibility:
1. Create an environment and install toga-core, toga-<your_platform>, and toga-demo all at 0.4.8, then install the local dev version of Travertino (can't be done in the same line because its version won't read as high enough to satisfy requirements), and run the demo to see if it explodes. I've added a CI target that essentially does this (albeit only with the dummy backend).
2. Create a second repo clone and set it to the 0.4.8 tag, so that you have 0.4.8's test suite. Edit its tox.ini to install 0.4.8 from PyPI and Travertino from your dev repo, and to run `pytest` with `-Wignore` to suppress warnings, then run its core test suite:
```ini
commands =
    python -m pip install 'toga_core[dev]==0.4.8' toga_dummy==0.4.8
    python -m pip install -e 'path/to/local/travertino'
    python -m pytest -Wignore {posargs:-vv --color yes}
```

The second is arguably more thorough, but also harder to interpret; some of its tests are always going to fail even if nothing's functionally broken, due to assertions about implementation details that have changed. I also tried running the testbed — which involved editing development Travertino to lie about its version — and unfortunately, it hangs indefinitely somewhere in the widget properties. Not sure why.

So... in short, this updated state may or may not be 100% perfect, but it's at least *primarily* functional, and runs the demo and a handful of examples I've tried without incident, which is, unfortunately, better than has been the case for the last five months. And the shim's too-wide net has been narrowed, so it shouldn't be masking unrelated errors.

## PR Checklist:

- [x] All new features have been tested
- [x] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

### CODE REVIEW

**Comment by HalfWhitt** [comment]
I've also altered the wording of some deprecation warnings; it was confusing that they referred to themselves as "this error" when they're only warnings.

**Comment by johnzhou721** [comment]
> and I'm honestly surprised no one's come asking about it

FWIW:  I was going to ask about this while doing the Qt backend, since I hit tons of things caught by this, but I thought experienced devs won't make those errors, and asking about it might make me seem that I make tons of stupid mistakes.

**Comment by HalfWhitt** [comment]
> > and I'm honestly surprised no one's come asking about it
> 
> FWIW: I was going to ask about this while doing the Qt backend, since I hit tons of things caught by this, but I thought experienced devs won't make those errors, and asking about it might make me seem that I make tons of stupid mistakes.

I meant specifically that people hadn't come asking about why their apps broke when they did a reinstall from requirements and wound up with Toga 0.4.8 and Travertino 0.5.x. But the other side of it (this warning being generated when it's not relevant, and obscuring other errors) is certainly a problem too, and I wish I'd realized earlier how troublesome it could be. Please feel free to bring up when things like this don't seem to be behaving right. : )

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -190,6 +190,40 @@ jobs:
...
+    - name: Test
+      run: |
+        export TOGA_BACKEND=toga_dummy
+        # Install pytest (and PIL) so dummy is happy, see #2290
+        python -m pip install toga_core==0.4.8 toga_dummy==0.4.8 pytest pillow
+        # Install demo without dependencies to skip the backend, and overwrite the
+        # installed version of Travertino with the one being tested.
+        python -m pip install --no-dependencies toga_demo==0.4.8 ./$(ls dist/travertino-*.whl)
+        # Nothing fancy, just a quick sanity check to make sure it runs without error.
+        python -m toga_demo
```
Comment (freakboy3742) [inline .github/workflows/ci.yml]: I'm not wild about this as a test. It took a while for me to even work out what was happening, because the CI log doesn't make it easy to see what is happening - toga_demo is crashing (which is expected), but there's no crash log in CI. And if it *doesn't* crash... you'll have a running app, so CI will lock up.

I think we might need to add a single `test_compat.py` test file, and run that test as a "standalone" backwards compatibility test suite that validates the edge cases we're expecting - raising warnings, an exception on Pack.apply, etc.

File: travertino/src/travertino/style.py [hunk 2]
```diff
@@ -72,29 +72,66 @@ def _applicator(self, value):
...
+                    "assigned an applicator."
                 )
-            ######################################################################
-            # End backwards compatibility
-            ######################################################################
+
+                from importlib.metadata import PackageNotFoundError, version
+
+                try:
+                    major, minor, patch = version("toga_core").split(".", maxsplit=2)
```
Comment (freakboy3742) [inline travertino/src/travertino/style.py]: `packaging.version.Version("0.4.8")` returns an object that is comparable with other `Version` objects.

File: travertino/src/travertino/properties/validated.py [hunk 3]
```diff
@@ -57,7 +57,22 @@ def __set__(self, style, value):
...
         setattr(style, f"_{self.name}", value)
         if value != current:
-            style.apply(self.name)
+            ######################################################################
+            # 08-2025: Backwards compatibility for Toga < 0.5.0
+            ######################################################################
+            try:
+                style.apply(self.name)
+            except TypeError as exc:  # pragma: no cover
+                if str(exc) == (
```
Comment (freakboy3742) [inline travertino/src/travertino/properties/validated.py]: I'm not wild about literal checks of exception text - these messages can (and do) change between Python versions.

I'd suggest one of two approaches here:
1. Use `importlib.metadata.metadata("toga-core")["version"]` to get the installed version of `toga-core`, and confirm if it's < 0.5.0 (using `packaging.version.Version`)
2. Use `inspect.getfullargspec(Pack.apply).args` to check if Pack.apply has three args (`["self", "prop", "value"]`); if it does, invoke as appropriate; if not, raise the original exception

File: travertino/src/travertino/properties/validated.py [hunk 4]
```diff
@@ -67,7 +82,22 @@ def __delete__(self, style):
...
         else:
             if current != self.initial:
-                style.apply(self.name)
+                ######################################################################
+                # 08-2025: Backwards compatibility for Toga < 0.5.0
+                ######################################################################
+                try:
+                    style.apply(self.name)
+                except TypeError as exc:  # pragma: no cover
+                    if str(exc) == (
```
Comment (freakboy3742) [inline travertino/src/travertino/properties/validated.py]: See notes above about exception literals.

File: travertino/src/travertino/style.py [hunk 5]
```diff
@@ -72,29 +72,66 @@ def _applicator(self, value):
...
+            # more specific RuntimeError.
+
+            try:
+                try:
+                    self.apply()
+                except TypeError as type_exc:  # pragma: no cover
+                    if str(type_exc) == (
+                        "Pack.apply() missing 2 required positional arguments: "
+                        "'prop' and 'value'"
+                    ):
```
Comment (freakboy3742) [inline travertino/src/travertino/style.py]: See above comments about exception literals.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (HalfWhitt) [reply .github/workflows/ci.yml]: Crap, I hadn't noticed it crashed. Why doesn't that give a failure exit code? Thank you for pointing it out.

However, no, that's *not* expected. That's what happens *without* the changes in this PR. It looks like in the command where I attempt to install the development version of Travertino that was previously built, the `ls()` command is instead finding the 0.5.2 version that was downloaded in the previous step. I hadn't even thought of that possibility.

I'm running the "test" using the dummy backend; it doesn't have an event loop, so if it runs successfully, it shouldn't lock up. It should simply go through the motions of running the app startup method and calculating layouts, spitting out a lot of deprecation warnings along the way, and then exit; this is what it does for me locally (with Toga / Toga demo 0.4.8 and this PR's version of Travertino).

All that sad... yes, I realize it's definitely not a robust test. All it verifies (or should verify, if I were to fix which version of Travertino is installed, and get an exception to count as test failure) is that the shims handle things well enough that the demo's startup app and layout mechanisms can run to completion without throwing an uncaught exception.

Where exactly would I put such a compatibility test? If I put it in the tests folder, it'll get sucked up by pytest, right? So I guess I just put it elsewhere in Travertino, and then call it manually in CI...?

Comment (HalfWhitt) [reply .github/workflows/ci.yml]: Question: are development wheels always versioned 0.1.dev[string of characters]? That is, is there any chance the 0.1. part can change, or can I count on that and hardcode it into a glob match?

Comment (freakboy3742) [reply .github/workflows/ci.yml]: > However, no, that's _not_ expected. That's what happens _without_ the changes in this PR. It looks like in the command where I attempt to install the development version of Travertino that was previously built, the `ls()` command is instead finding the 0.5.2 version that was downloaded in the previous step. I hadn't even thought of that possibility.

Are you sure about that? The [last step of the CI log before toga-demo starts](https://github.com/beeware/toga/actions/runs/16849970527/job/47734708682?pr=3683#step:5:92) indicates that 0.5.3dev45+gac96e234d was installed, which would be consistent with the packages that have been build with the CI pass.
 
> I'm running the "test" using the dummy backend; it doesn't have an event loop, so if it runs successfully, it shouldn't lock up.

Using Toga-dummy is a good approach to avoid the lockup... except you're not running on toga-dummy. When you `pip install toga-demo`, that installs toga as a dependency, which installs toga-cocoa (or whatever backend is local), and that will take priority over dummy. 
 
> Where exactly would I put such a compatibility test? If I put it in the tests folder, it'll get sucked up by pytest, right? So I guess I just put it elsewhere in Travertino, and then call it manually in CI...?

Yeah - another test directory (`travertino/compat`?) with an independent pytest pass makes sense to me. 

> Question: are development wheels always versioned 0.1.dev[string of characters]? That is, is there any chance the 0.1. part can change, or can I count on that and hardcode it into a glob match?

The wheel version comes from git tags; if there are no tags, it falls back to "0.1" as the initial release. If you're seeing 0.1.dev[hash], that would normally indicate you haven't done a `git pull --tags` into the checkout. Again, CI doesn't show any 0.1 naming.

Comment (HalfWhitt) [reply .github/workflows/ci.yml]: > Are you sure about that? The [last step of the CI log before toga-demo starts](https://github.com/beeware/toga/actions/runs/16849970527/job/47734708682?pr=3683#step:5:92) indicates that 0.5.3dev45+gac96e234d was installed, which would be consistent with the packages that have been build with the CI pass.

> The wheel version comes from git tags; if there are no tags, it falls back to "0.1" as the initial release. If you're seeing 0.1.dev[hash], that would normally indicate you haven't done a `git pull --tags` into the checkout. Again, CI doesn't show any 0.1 naming.

> Using Toga-dummy is a good approach to avoid the lockup... except you're not running on toga-dummy. When you `pip install toga-demo`, that installs toga as a dependency, which installs toga-cocoa (or whatever backend is local), and that will take priority over dummy.

Welp... there I go not looking at the relevant things again 🤦


> > Where exactly would I put such a compatibility test? If I put it in the tests folder, it'll get sucked up by pytest, right? So I guess I just put it elsewhere in Travertino, and then call it manually in CI...?
> 
> Yeah - another test directory (`travertino/compat`?) with an independent pytest pass makes sense to me.

👍

Comment (HalfWhitt) [reply .github/workflows/ci.yml]: Holy crap... I just figured out why I've never been able to access any tags in any of my local repo clones (and why the local testing versions are all 0.1.devxxx). I've never been able to fetch tags because they're not on *my* fork, and I don't think there's any way to tell GitHub to sync them.

I've added upstream as a remote and pulled the tags from there (and also pushed them to my fork on GitHub). After all this time, I can finally use tags! This should absolutely go in the contribution guide.

File: travertino/src/travertino/properties/validated.py (hunk 3 above)
Comment (HalfWhitt) [reply travertino/src/travertino/properties/validated.py:66]: > Use `importlib.metadata.metadata("toga-core")["version"]`

Does this have any advantage over the built-in `version("toga_core")`?

Comment (HalfWhitt) [reply travertino/src/travertino/properties/validated.py]: That makes sense, though I think to do it right, that would have to be an `if`/`else`, rather than a `try`/`except`. You could get a `TypeError` during `apply()` that's *not* due to the version mismatch, even if the versions are in fact mismatched. If we can't tell the difference in the exception itself, we need to preemptively decide how to call the method.

Comment (freakboy3742) [reply travertino/src/travertino/properties/validated.py]: I think the try-catch is still needed though. We want the "normal" path to be "just call `apply()`"; but, if an error occurs, try to diagnose if version mismatch is the likely cause, and if it is, handle appropriately.

Comment (HalfWhitt) [reply travertino/src/travertino/properties/validated.py]: Okay... I guess the only time it could mask an unrelated error is the there's already a version mismatch too, come to think of it.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (HalfWhitt) [reply .github/workflows/ci.yml]: I'm still working on the more robust testing setup, but I've been scratching my head trying to figure out how the demo test-run failed in CI... it was Toga 0.4.8, Travertino 0.5.2, and the dummy backend set by environment variable, which all works fine together locally.

Turns out, it didn't fail. That's just the sheer number of deprecation warnings it gives when it runs successfully to completion! 😂

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -125,9 +125,7 @@ jobs:
       run: |
         # The $(ls ...) shell expansion is done in the GitHub environment; the value of
         # TOGA_INSTALL_COMMAND will be a literal string without any shell expansions to
-        # perform. We don't *explicitly* install the Travertino wheel; we provide the
-        # dist dir as a source of wheels, and rely on pip to resolve the explicit
-        # version match to the Travertino wheel in that folder.
```
Comment (HalfWhitt) [inline .github/workflows/ci.yml:130]: Also removed this line, which is no longer true but I somehow completely missed it when I did #3655.

**Comment by HalfWhitt** [comment]
- Changed how the errors are matched, inspecting the Toga version instead of the error string
- Added tests to confirm that the shims aren't hiding unrelated errors
- Added a test environment to tox and CI to check Travertino with Toga 0.4.8

The last of these is better than nothing, but I'm pretty sure it could be more robust. I'll try and come back to this with fresh eyes later this week, but for now, it's at least all passing.

File: travertino/compat/test_compat.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+        _ = Pack(fake_name=False)
+
+
+def test_demo():
+    """The demo app can start up (with the demo backend).
+
+    This doesn't actually generate any deprecation warnings, because they've already
+    been emitted when importing Toga. But this confirms that, at the very least, the
+    demo app can run through its startup method without generating any uncaught
+    exceptions.
```
Comment (freakboy3742) [inline travertino/compat/test_compat.py:76]: Worth mentioning this is non-blocking because it's running against the dummy backend.

File: travertino/src/travertino/style.py [hunk 8]
```diff
@@ -5,12 +5,19 @@
...
+from .compat import _toga_lt_5
 from .properties.shorthand import directional_property
 from .properties.validated import validated_property
 
 # Make sure deprecation warnings are shown by default
 filterwarnings("default", category=DeprecationWarning)
 
+_DEPRECATION_MSG = (
+    "You're probably seeing this because you've updated Travertino to 0.5.x but are "
+    "still using Toga <= 0.4.8; to fix, either update Toga to >= 0.5.0, or pin "
```
Comment (freakboy3742) [inline travertino/src/travertino/style.py]: "still" is mildly prejorative here.
```suggestion
    "using Toga <= 0.4.8; to fix, either update Toga to >= 0.5.0, or pin "
```

Comment (HalfWhitt) [reply travertino/src/travertino/style.py]: Ah, fair enough. I only meant to contrast in time, but I see it how it does sound that way.

**Comment by HalfWhitt** [comment]
@freakboy3742 I still suspect the tests I've added aren't comprehensive enough, but either I'm wrong or I could use a nudge from someone who's not quite as close to it. The first three tests (`test_reapply`, `test_validated_set_delete`, and `test_invalid_property_in_init`) hit all the specific shims that I've added (that aren't already tested in the "real" Travertino test suite), and the final one (`test_demo`) is essentially a very low-stakes integration test, verifying that the pieces all run together without error.

So that's the lowest and the highest level. There are any number of things in between I could test, but I'd essentially be recreating (some portion of) the standard test suite(s). Anything seem like relevant hot spots to you?

One option: I could test that *importing Toga* a) works and b) generates deprecation warnings. But then that would mean the other tests would fail if it weren't run first. (There's always `importlib.reload` to check within a test, but that won't re-generate the same deprecation warnings.)

File: .github/workflows/ci.yml [hunk 9]
```diff
@@ -190,6 +188,41 @@ jobs:
...
+    steps:
+    - name: Checkout
+      uses: actions/checkout@v4.2.2
+      with:
+        fetch-depth: 0
+
+    - name: Set up Python 3.10
+      uses: actions/setup-python@v5.6.0
+      with:
+        python-version: "3.10"
```
Comment (freakboy3742) [inline .github/workflows/ci.yml]: This should probably be `min_python_version` so we don't have to update it.

File: .github/workflows/ci.yml [hunk 10]
```diff
@@ -190,6 +188,41 @@ jobs:
...
+    - name: Get Package
+      uses: actions/download-artifact@v4.3.0
+      with:
+        pattern: ${{ format('{0}-*', needs.package.outputs.artifact-basename) }}
+        merge-multiple: true
+        path: dist
+
+    - name: Test
+      run: |
+        export TRAVERINO_INSTALL_COMMAND="python -m pip install ../$(ls dist/travertino-*.whl)"
```
Comment (freakboy3742) [inline .github/workflows/ci.yml]: Typo: 
```suggestion
        export TRAVERTINO_INSTALL_COMMAND="python -m pip install ../$(ls dist/travertino-*.whl)"
```

**Comment by HalfWhitt** [comment]
> If there are extra things to test here, I'm not seeing them. You're covering all the use cases I can think of.
> 
> I did pick up a couple of minor CI config issues; but otherwise, I'm happy to land this.

Thanks for the fixes! I guess it just felt like not much testing, but sometimes there's not much to test.
